File: src/main/java/net/orandja/vw/mixin/ExplosionMixin.java
package net.orandja.vw.mixin;

import lombok.Getter;
import lombok.Setter;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import net.minecraft.world.explosion.Explosion;
import net.orandja.vw.logic.ProtectBlock;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.Shadow;
import org.spongepowered.asm.mixin.injection.At;
import org.spongepowered.asm.mixin.injection.Inject;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;

import java.util.List;

@Mixin(Explosion.class)
abstract class ExplosionMixin implements ProtectBlock {
    @Shadow @Getter World world;

    @Shadow @Getter @Setter List<BlockPos> affectedBlocks;

    @Inject(method = "affectWorld", at = @At("HEAD"))
    void affectWorld(boolean particles, CallbackInfo info) {
        onExplosion();
    }

}

File: src/main/java_old/kotlin_old/kmixin/ExplosionMixin.kt
@file:Mixin(Explosion::class)
@file:JvmName("ExplosionMixin")

package net.orandja.vw.kmixin

import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import net.minecraft.world.explosion.Explosion
import net.orandja.vw.mods.ProtectBlock
import org.spongepowered.asm.mixin.Mixin
import org.spongepowered.asm.mixin.Shadow
import org.spongepowered.asm.mixin.injection.At
import org.spongepowered.asm.mixin.injection.Inject
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo

@Mixin(Explosion::class)
abstract class ExplosionMixin {
    @Shadow
    var world: World? = null

    @Shadow
    var affectedBlocks: List<BlockPos>? = null

    @Inject(method = ["affectWorld"], at = [At("HEAD")])
    fun affectWorld(particles: Boolean, info: CallbackInfo) {
        val newBlocks: List<BlockPos> = affectedBlocks!!.filterNot { pos: BlockPos ->
            ProtectBlock.canExplode(world!!, pos!!)
        }
        (this.affectedBlocks as ArrayList<BlockPos>).apply {
            this.clear()
            this.addAll(newBlocks)
        }
    }

}